src/utils.py

```diff
@@ -1860,11 +1861,62 @@ def sample_subsets(universe: Sequence[_T], num_samples: int, min_set_size: int,
 def create_ground_atom_dataset(
         trajectories: Sequence[LowLevelTrajectory],
         predicates: Set[Predicate]) -> List[GroundAtomTrajectory]:
-    """Apply all predicates to all trajectories in the dataset."""
-    ground_atom_dataset = []
-    for traj in trajectories:
-        atoms = [abstract(s, predicates) for s in traj.states]
-        ground_atom_dataset.append((traj, atoms))
+    """Generate or load a GroundAtomTrajectory by applying all predicates to
```

Comment:
this is a really good point, agreed! Any thoughts on which function we should add this to (I can't seem to find a good place for it)? My current inclination is that the right place is in `nsrt_learning_main.py`: instead of always creating a new dataset on line 41, we can instead load depending on the flag (here, we can be assured that the trajectories and predicates correspond to the offline dataset, I think)?

Reply:
that would be a little better, but still has some of the same risks -- for example, if we run online learning, then `nsrt_learning_main.py` would get used multiple times, with different trajectories each time.

another idea (which is gonna look not pretty, but maybe is the safest bet): add an optional kwarg to `learn_nsrts_from_data` called `ground_atom_dataset`, which if provided, is used instead of running STEP 1. then do the saving and loading inside of `NSRTLearningApproach._learn_nsrts`, where importantly, we can use the `online_learning_cycle` in the cache file name.